### PR TITLE
Add takeP with tests and benchmarks

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

File: src/Streamly/Internal/Data/Parser/ParserD.hs [hunk 1]
```diff
@@ -814,6 +815,48 @@ spanByRolling eq f1 f2 =
 -- nested parsers
 -------------------------------------------------------------------------------
 
+-- | See 'Streamly.Internal.Data.Parser.takeP'.
+--
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Parser/ParserD.hs:819]: /Internal/

File: src/Streamly/Internal/Data/Parser.hs [hunk 2]
```diff
@@ -852,6 +853,21 @@ longest p1 p2 = K.toParserK $ D.longest (K.fromParserK p1) (K.fromParserK p2)
...
+--
+-- * Stops - when the collecting parser stops.
+-- * Fails - when the collecting parser fails.
+--
+-- >>> Stream.parse (Parser.takeP 4 (Parser.takeEQ 2 Fold.toList)) $ Stream.fromList [1, 2, 3, 4, 5]
+-- [1, 2]
+--
+-- >>> Stream.parse (Parser.takeP 4 (Parser.takeEQ 5 Fold.toList)) $ Stream.fromList [1, 2, 3, 4, 5]
+-- *** Exception: ParseError "takeEQ: Expecting exactly 5 elements, input terminated on 4"
+--
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Parser.hs:866]: /Internal/

File: src/Streamly/Internal/Data/Parser.hs [hunk 3]
```diff
@@ -852,6 +853,21 @@ longest p1 p2 = K.toParserK $ D.longest (K.fromParserK p1) (K.fromParserK p2)
 lookAhead :: MonadCatch m => Parser m a b -> Parser m a b
 lookAhead p = K.toParserK $ D.lookAhead $ K.fromParserK p
 
+-- | Takes at-least @n@ input elements.
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Parser.hs:856]: Shouldn't it be like `take` i.e. `at most n`?

File: src/Streamly/Internal/Data/Parser/ParserD.hs [hunk 4]
```diff
@@ -817,45 +817,53 @@ spanByRolling eq f1 f2 =
...
-        return
-            $ case res of
-                  IPartial s -> IPartial (Tuple' 0 s)
-                  IDone b -> IDone b
-                  IError e -> IError e
-
-    step (Tuple' cnt r) a
-        | cnt < lim = do
-            res <- fstep r a
+            res <- pinitial
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Parser/ParserD.hs]: Use 4 space indentation.

File: src/Streamly/Internal/Data/Parser/ParserD.hs [hunk 5]
```diff
@@ -817,45 +817,53 @@ spanByRolling eq f1 f2 =
...
+                    if cnt1 < lim
+                    then return $ Partial 0 $ Tuple' cnt1 s
+                    else Done 0 <$> pextract s
+                Continue 0 s -> do
+                    assert (cnt1 >= 0) (return ())
+                    if cnt1 < lim
+                    then return $ Continue 0 $ Tuple' cnt1 s
+                    -- XXX This should error out?
+                    -- If designed properly, this will probably error out.
+                    -- "pextract" should error out
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Parser/ParserD.hs]: This is a tricky case, we have the following options:

1. `Done 0` with extract as you have written
2. `Done n`, will require buffering elements
3. Use a backtracking fold and not a parser, once we have backtracking in folds

**Review by harendra-kumar — APPROVED** [review]

**Pull request merged**